### PR TITLE
feat(instancedata): cleanup charmprofiles and watch instance data

### PR DESCRIPTION
**Pull request by nvinuesa**
After this patch there will be no more used methods on machine related
to instance data on the legacy state.

This patch cleans up and wires CharmProfiles() (now called
`AppliedLXDProfiles()` on the new domain) on the missing call sites.
Also, the InstanceDataWatcher was cleaned up.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

LXD profiles are currently broken on main (until the migration for units and apps is finished) so we cannot QA using the CI tests deploy-lxd-profiles.
_Note: I haven't spent a huge amount of time trying to fix it (it's out of scope because it touches other domains) but I think it fails because the instance mutater is not triggering the call to write the profiles, maybe due to some of the other watchers (units/apps?) not being fully wired._
Bootstrap and deploy should work with no errors though:
```
juju bootstrap lxd c
juju add-model m
juju deploy ubuntu
```
Check status and logs for any unexpected error.

## Links

**Jira card:** JUJU-6703

### CODE REVIEW

**Review by hmlanigan — COMMENTED** [review]
A few nits and some questions. In general the changes look good.

Technically  the cli commands in QA steps do not return an error. However there are errors `juju status` and in the logs:
```
unit-lxd-profile-without-devices-0: 19:01:15 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: required charm profile on machine not found
```
and the workload is never set to active in status. 

Please update the PR with more detail on expected result and specific of how profiles are broken, or link to the bug. It's unclear what is expected or not.

File: apiserver/facades/client/client/service.go [hunk 1]
```diff
@@ -43,4 +43,6 @@ type MachineService interface {
 	// InstanceIDAndName returns the cloud specific instance ID and display name for
 	// this machine.
 	InstanceIDAndName(ctx context.Context, machineUUID string) (string, string, error)
+	// LXDProfiles returns the names of the LXD profiles on the machine.
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/facades/client/client/service.go]: Fixed.

File: apiserver/facades/agent/instancemutater/instancemutater.go [hunk 2]
```diff
@@ -49,6 +50,10 @@ type MachineService interface {
 	// lxd_profile table for the given machine. This method will overwrite the list
 	// of profiles for the given machine without any checks.
 	SetAppliedLXDProfileNames(ctx context.Context, mUUID string, profileNames []string) error
+
+	// WatchLXDProfiles returns a NotifyWatcher that is subscribed to the changes in
+	// the machine_cloud_instance table in the model, for the given machine UUID.
+	WatchLXDProfiles(ctx context.Context, machineUUID string) (watcher.NotifyWatcher, error)
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/facades/agent/instancemutater/instancemutater.go:55]: Nice catch! fixed.

**Review by jack-w-shaw — APPROVED** [review]
LGTM!

There is another method in state you can remove according to static analysis

**Review by SimonRichardson — APPROVED** [review]
We need to be sure that we're correctly handling the errors esp. the NotProvisioned error.

**Comment by hmlanigan** [comment]
@nvinuesa If you tell me that deploy should be successful, I assume that `juju status` shows active/idle and is green for the new application. The deploy command returning is insufficient as that only means there were no errors in the provided arguments. 90% of deploy happens after the command returns. 

Based on your knowledge of what is broken, what is the expectation for `juju status` in this scenario? It should be in the QA steps. 

I understand that lxd profiles will be fixed later once more of application and charm are done. I'd like to have an idea if this change is making it worse or not.

**Comment by nvinuesa** [comment]
@hmlanigan as I said earlier (offline and on this PR) I did spend some time trying to fix the instancemutater and lxdprofilewatcher but I didn't go any further because it was timeboxed. However I can confirm that the profiles are never set because the changes expected from units and applications are never observed and therefore the call to `SetAppliedLXDProfileNames` doesn't happen.
This is the reason why I didn't add the deploy with a charm with lxd profiles on the QA steps. If you were to deploy a charm with lxd profiles then it wouldn't reach an active status because it looks for profiles that are not retrieved from the model db. This behavior was already broken on main (using the full implem from legacy state). 

As per the status expectations, it is part of the QA to "Check status and logs for any unexpected error.", but of course, since we are deploying a charm without lxd profiles no errors are expected there.

We can continue discussing this subject if the answer is not satisfying (and add a new epic item on the next cycle?).

**Pull request merged**